knimin/lib/data_access.py

```diff
@@ -2450,6 +2452,866 @@ def set_deposited_ebi(self):
+        """
+        with TRN:
+            self._sample_plate_exists(plate_id)
+            self._clear_sample_plate_layout(plate_id)
+
+            sql = """DELETE FROM pm.sample_plate_study
+                     WHERE sample_plate_id = %s"""
+            TRN.add(sql, [plate_id])
+            sql = """DELETE FROM pm.sample_plate
+                     WHERE sample_plate_id = %s"""
+            TRN.add(sql, [plate_id])
+            TRN.execute()
+
+    def get_property_options(self, prop):
+        """Retrieves a list of available options for a property
+
+        Parameters
+        ----------
+        prop : str
+            Property name, i.e., name of a table under schema "pm"
+
+        Returns
+        -------
+        list of dict
+            {id : int, name : str, notes : str}
+            ID, name and notes for each option
+        """
+        with TRN:
+            sql = """SELECT {} AS id, name, notes
+                     FROM {}
+                     ORDER BY {}"""
+            TRN.add(sql.format(prop + '_id',
+                               'pm.' + prop,
+                               prop + '_id'))
+            return [dict(x) for x in TRN.execute_fetchindex()]
+
+    def get_plate_types(self):
+        """Gets all available plate types
+
+        Returns
+        -------
+        list of dict
+            {id : int, name : str, cols : int, rows : int, notes : str}
+            ID, name, notes, and numbers of columns and rows of each plate type
+        """
+        with TRN:
+            sql = """SELECT plate_type_id AS id, name, notes, cols, rows
+                     FROM pm.plate_type
+                     ORDER BY plate_type_id"""
+            TRN.add(sql)
+            return [dict(x) for x in TRN.execute_fetchindex()]
+
+    def read_plate_type(self, plate_type_id):
+        """Returns the information the given plate type
+
+        Parameters
+        ----------
+        plate_type_id: int
+            The id of the plate type
+
+        Returns
+        -------
+        DictCursor
+            The information of the plate type
+        """
+        with TRN:
+            sql = """SELECT plate_type_id, name, cols, rows, notes
+                     FROM pm.plate_type
+                     WHERE plate_type_id = %s"""
+            TRN.add(sql, [plate_type_id])
+            res = TRN.execute_fetchindex()
+            if not res:
+                raise ValueError("Plate type %s doesn't exist" % plate_type_id)
+            # Magic number 0: there is only one result in the query
+            return res[0]
+
+    def get_emails(self):
+        """Gets all available emails
+
+        Returns
+        -------
+        list of str
+            Sorted list of emails
+        """
+        with TRN:
+            sql = """SELECT email
+                     FROM ag.labadmin_users
+                     ORDER BY email"""
+            TRN.add(sql)
+            return TRN.execute_fetchflatten()
+
+    def get_sample_plate_ids(self):
```

Comment:
No `WHERE` clause? This will yield a large number of plates, which will only get larger over time -- if this is for a UI element, then it may make sense to paginate